spec/truffle/interop/key_info_spec.rb

```diff
@@ -0,0 +1,201 @@
+# redistribute it and/or modify it under the terms of the:
+# 
+# Eclipse Public License version 1.0
+# GNU General Public License version 2
+# GNU Lesser General Public License version 2.1
+
+require_relative '../../ruby/spec_helper'
+
+class KeyInfoFixture
+  
+  def initialize
+    @foo = 14
+  end
+
+  attr_reader :ro
+  attr_accessor :rw
+  attr_writer :wo
+
+end
+
+describe "Truffle::Interop.key_info" do
+  describe "for a Hash with String keys" do
+    
+    before :each do
+      @hash = {'a' => 1, 'b' => 2, 'c' => 3}
+      @hash.instance_variable_set(:@foo, 14)
+    end
+    
+    it "returns :existing for all keys" do
+      @hash.keys.each do |k|
+        Truffle::Interop.key_info(@hash, k).should include(:existing)
+      end
+    end
+    
+    it "returns :readable for all keys" do
+      @hash.keys.each do |k|
+        Truffle::Interop.key_info(@hash, k).should include(:readable)
+      end
+    end
+    
+    it "returns :writable for all keys" do
+      @hash.keys.each do |k|
+        Truffle::Interop.key_info(@hash, k).should include(:writable)
+      end
+    end
+    
+    it "does not return :writable for all keys if the hash is frozen" do
+      @hash.freeze
+      @hash.keys.each do |k|
+        Truffle::Interop.key_info(@hash, k).should_not include(:writable)
+      end
+    end
+    
+    it "does not return :invocable" do
+      @hash.keys.each do |k|
+        Truffle::Interop.key_info(@hash, k).should_not include(:invocable)
+      end
+    end
+    
+    it "does not return :internal for keys" do
+      @hash.keys.each do |k|
+        Truffle::Interop.key_info(@hash, k).should_not include(:internal)
+      end
+    end
+    
+    it "returns nothing for a missing key" do
+      Truffle::Interop.key_info(@hash, :key_not_in_key_info_hash).should == []
+    end
+    
+    it "returns :existing for an instance variable that exists" do
+      Truffle::Interop.key_info(@hash, :@foo).should include(:existing)
+    end
+
+    it "does not return :existing for an instance variable that does not exist" do
+      Truffle::Interop.key_info(@hash, :@bar).should_not include(:existing)
+    end
+    
+    it "returns :readable for an instance variable" do
+      Truffle::Interop.key_info(@hash, :@foo).should include(:readable)
+    end
+    
+    it "does not return :readable for an instance variable that does not exist" do
+      Truffle::Interop.key_info(@hash, :@bar).should_not include(:readable)
+    end
+    
+    it "does not return :writable for an instance variable if the hash is frozen" do
+      @hash.freeze
+      Truffle::Interop.key_info(@hash, :@foo).should_not include(:writable)
+    end
+    
+    it "returns :writable for an instance variable" do
+      Truffle::Interop.key_info(@hash, :@foo).should include(:writable)
+    end
+    
+    it "returns :writable for an instance variable that does not exist" do
+      Truffle::Interop.key_info(@hash, :@bar).should include(:writable)
```

Comment:
In such a case, :existing` is false as tested above. That makes sense.
But the documentation above contradicts it:
```
For all objects:
EXISTING is set if either READABLE or WRITABLE are set.
```
Maybe EXISTING is the same as READABLE?

Reply:
Yes, that's it.

Reply:
Actually `EXISTING` is the same as `READABLE` for `Hash`, but for other objects, `EXISTING` is `READABLE | WRITABLE` because of write-only setters.

Reply:
The JavaDoc mentions:
```
When the key does not exist (0. bit is zero), then all other bits must be zero as well.
```
Which seems strange, worth asking the Truffle team.